VERSIONING.md

```diff
@@ -0,0 +1,209 @@
+  * Modules will be used to encapsulate signals and components.
+    * Experimental modules still under active development will be versioned at
+      `v0` to imply the stability guarantee defined by
+      [semver](https://semver.org/spec/v2.0.0.html#spec-item-4).
+
+      > Major version zero (0.y.z) is for initial development. Anything MAY
+      > change at any time. The public API SHOULD NOT be considered stable.
+
+    * Mature modules that we guarantee a stable public API will be versioned
+      with a major version greater than `v0`.
+    * Experimental modules will start their versioning at `v0.0.0` and will
+      increment their minor version when backwards incompatible changes are
+      released and increment their patch version when backwards compatible
+      changes are released.
+    * All stable modules that use the same major version number will use the
+      same entire version number.
+      * Stable modules may be released with an incremented minor or patch
+        version even though that module has not been changed, but rather so
+        that it will remain at the same version as other stable modules that
+        did undergo change.
+      * When an experimental module becomes stable a new stable module version
+        will be released and will include this now stable module. The new
+        stable module version will be an increment of the minor version number
+        and will be applied to all existing stable modules as well as the newly
+        stable module being released.
+* Versioning of the associated [contrib
+  repository](https://github.com/open-telemetry/opentelemetry-go-contrib) of
+  this project will be idiomatic of a Go project using [Go
+  modules](https://github.com/golang/go/wiki/Modules).
+  * [Semantic import
+    versioning](https://github.com/golang/go/wiki/Modules#semantic-import-versioning)
+    will be used.
+    * Versions will comply with [semver 2.0](https://semver.org/spec/v2.0.0.html).
+    * If a module is version `v2` or higher, the
+      major version of the module must be included as a `/vN` at the end of the
+      module paths used in `go.mod` files (e.g., `module
+      go.opentelemetry.io/contrib/instrumentation/host/v2`, `require
+      go.opentelemetry.io/contrib/instrumentation/host/v2 v2.0.1`) and in the
+      package import path (e.g., `import
+      "go.opentelemetry.io/contrib/instrumentation/host/v2"`). This includes
+      the paths used in `go get` commands (e.g., `go get
+      go.opentelemetry.io/contrib/instrumentation/host/v2@v2.0.1`.  Note there
+      is both a `/v2` and a `@v2.0.1` in that example. One way to think about
+      it is that the module name now includes the `/v2`, so include `/v2`
+      whenever you are using the module name).
+    * If a module is version `v0` or `v1`, do not include the major version
+      in either the module path or the import path.
+  * In addition to public APIs, telemetry produced by stable instrumentation
```

Comment:
Do we need to define stable instrumentation?  E.g. can new labels be added?  Can the values of a label change?

Reply:
Yeah, I can take another try and nailing down what stable telemetry means for stable instrumentation. I'm open to suggestions if you have any. I'm thinking the telemetry can accrete but not reduce or change. So new labels can be added, but label values or keys cannot change.

Reply:
It depends on what we are trying to accomplish.  [Kubernetes, for example](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md#stability-classes), cares primarily about ingestion backwards-compatibility, rather than dashboard/alert backwards-compatibility.  Some metric backends are table-based, so adding (or removing) label keys isn't considered backwards-compatible for all metric backends. However, kubernetes doesn't require alert/dashboard stability, so label values can be added or removed, and metric values can be changed (e.g. by adding new streams).

I was involved with the kubernetes definition, so i'm biased :).

Label value stability is hard to define if the underlying thing it is measuring is allowed to change.  E.g. if an experimental processor is renamed, a metric about all processors with a processor_name label key would have a value that changes.

Reply:
@dashpole this is a great question!  I believe one of the goals of OpenCensus was to make it possible to add labels without breaking compatibility, to allow instrumentation to evolve and/or for users to configure which label dimensions they use. A frequent question we get is about the distinction between `UpDownSumObserver` and `ValueObserver`. 

One way to explain this distinction is that it supports automatic label removal without changing semantics. When removing labels from an UpDownSumObserver we sum the corresponding data points into a timeseries of the sum. When removing labels from a ValueObserver, we simply remove labels into a timeseries of reduced-dimension values.

The OTel-Go Metrics SDK has a "reducer processor" that implements label removal, for example, which is a mechanism OTel has to support OpenCensus views. Handling this "semantic removal of labels" in the OTel Collector requires more than just stripping labels, and is probably something the project will need for compatibility with "table-based" systems.

Reply:
@dashpole, if I understand you correctly, you're saying this part of [the otep]:

> Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard. This means that existing data may not be mutated or removed without a major version bump. Additional data may be added. This applies to spans, metrics, resources, attributes, events, and any other data types that OpenTelemetry emits.

Is unrealizable? Or, that I need to change the wording here to match that better?

Reply:
Unrealizable.  It looks like Josh, my TL, has already raised the same point in the OTEP: https://github.com/open-telemetry/oteps/pull/143#discussion_r541787167, which is probably a better place to have that discussion.  We can just wait for the outcome there and follow.  The wording here matches what is in the OTEP.

Reply:
With the merging of the linked OTEP we are going to stick with this language and adjust accordingly if the project language changes.